icu4c/source/i18n/collationbuilder.cpp

```diff
@@ -1128,6 +1139,11 @@ CollationBuilder::addOnlyClosure(const UnicodeString &nfdPrefix, const UnicodeSt
             UnicodeString str = stringIter.next();
             if(str.isBogus()) { break; }
             if(ignoreString(str, errorCode) || str == nfdString) { continue; }
+            if (loop++ > kClosureLoopSoftLimit) {
+                // To avoid hang as in ICU-22517, return with error.
+                errorCode = U_USELESS_COLLATOR_ERROR;
```

Comment:
This one is documented as “Collator is options only and no base is specified” which does not fit here.

I suggest we use [U_INPUT_TOO_LONG_ERROR](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/utypes_8h.html#a3343c1c8a8377277046774691c98d78caf9ce1d39861c5f6dff9935eaea228434).